Integrate with the Referrer Policy spec, part 2 of n

domenic opened the pull request:
This makes getting the referrer policy a first-class citizen of the
environment settings object algorithm set. WorkerGlobalScopes and
Documents both get referrer policies, and the process by which the
header or meta element set these is spelled out explicitly. This
includes tricky cases like navigation and `<iframe>`s, which the Referrer
Policy spec currently handles with (somewhat incomplete and incorrect)
monkeypatches.

Closes the following Referrer Policy issues:

- https://github.com/w3c/webappsec-referrer-policy/issues/30
- https://github.com/w3c/webappsec-referrer-policy/issues/32

and is part of the larger effort at https://github.com/w3c/webappsec-referrer-policy/issues/43

/cc @bzbarsky @jeisinger @estark37 @mikewest 

I think there are two important questions outstanding before this is mergeable:

- Is the "copying" of referrerpolicy for various nested browsing contexts correct? https://github.com/w3c/webappsec-referrer-policy/issues/30, https://github.com/w3c/webappsec-referrer-policy/issues/31, and https://github.com/w3c/webappsec-referrer-policy/issues/32 all touch on that. I assumed here the desire was to snapshot the creator document's referrer policy at new-document creation time. This leads to the potential issue pointed out by @bzbarsky in https://github.com/w3c/webappsec-referrer-policy/issues/31. An alternative is to have the new document's referrer policy point upward to the creator document, unless overridden by a meta tag; this seems messier though.

- I did not understand the intent of https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-delivery-implicit-workers. There is no such thing as "the Worker's incumbent settings object". If I were to interpret it as "the incumbent settings object at the time the Worker constructor is running," then on the face of it, what that algorithm says is that doing `new Worker()` will override your document's referrer policy and reset it to `"no-referrer-when-downgrade"` (i.e., this is a third way to set your document-wide policy, in addition to the HTTP header and the meta element). That seems clearly wrong. What was the intent of that section?

Also, @bzbarsky @annevk, I was a bit unsure how to reify the "local scheme nested browsing contexts inherit". I think I did it, but that section could use careful review. I don't think the navigate algorithm is very explicit about local schemes at the moment; it might be my patch is the best we can do until more navigation cleanup happens.

This should allow the following sections of Referrer Policy to become more informative and get trimmed down:

- https://w3c.github.io/webappsec-referrer-policy/#integration-with-html
- https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-delivery-implicit
- https://w3c.github.io/webappsec-referrer-policy/#terms (definition of referrer policy no longer needs to patch environment settings object)
- https://w3c.github.io/webappsec-referrer-policy/#set-referrer-policy (can be removed)

estark37 commented:
> I assumed here the desire was to snapshot the creator document's referrer policy at new-document creation time.

That's what I would expect also.

@annevk @bzbarsky have you had a chance to look at this PR yet?

bzbarsky commented:
I'm generally not terribly interested in the details here; I just filed the issues because I was looking at some code in Gecko that dealt with referrer policies and there were obvious problems with that code, so I looked at the spec and the spec was problematic too.  Oh, and there are serious problems with the wpt test coverage, leading to minor attempts like https://github.com/w3c/web-platform-tests/pull/2851 to fix that.  The total black hole there does not help me care about this topic, obviously.  ;)

One other thing for the future: these pull requests for the HTML spec are _much_ easier to review when I have an actual generated spec to look at instead of having to reverse-engineer it from the diff.

Anyway, on the substance... What Gecko does, I believe, is store the referrer policy in the thing that gets aliased when origin needs to alias, but does snapshot it at document creation time, copying it from that aliased thing.  Note that this applies to more than just nested browsing contexts; for example I'm pretty sure we copy the referrer policy when doing `window.open()` (no url).  I have no idea what other browsers do, and there is no test coverage for this afaict.  Looks to me like this spec proposal does do that, at first glance.

I don't know that "local URL" is the right thing to be using.  That includes "filesystem"; do we really want to inherit referrer policies for those?

The bit about "Otherwise (e.g., if the navigation was to a <span>URL</span> with a <span>local scheme</span>)" doesn't make sense to me either.  If I navigate to a blob: is there not a `response`? In what actual situations is there no `response`?  If I have a document with referrer policy A which creates a subframe, then a document with referrer policy B navigates that subframe to a blob:, is the referrer policy of the result A or B?  It should be B.  I think this spec proposal makes it A.

One last thing is what should happen with history traversal.  This is currently pretty broken anyway because https://html.spec.whatwg.org/multipage/browsers.html#history-traversal step 1 uses "source browsing context", but the relevant scenario is this one:

1. A document D1 with referrer policy P1 loads URL U1 in a subframe.
2. The subframe is then navigated to a different URL U2.
3. The referrer policy of D1 is changed to P2.
4. The user clicks the back button.  Which referrer policy is used when determining the referrer for the navigation to U1?  Is it P1 or P2?  Gecko uses P1; we actually capture the referrer and referrer policy used for the load in the history entry that results from loading U1, so we can replay the same exact fetch later.  Again, I have no idea what other browsers do and there is no test coverage.

domenic commented:
> The bit about "Otherwise (e.g., if the navigation was to a URL with a local scheme)" doesn't make sense to me either. If I navigate to a blob: is there not a response? In what actual situations is there no response? If I have a document with referrer policy A which creates a subframe, then a document with referrer policy B navigates that subframe to a blob:, is the referrer policy of the result A or B? It should be B. I think this spec proposal makes it A.

@annevk, could you help figuring this out? As I said in my OP, it might be that we need to clean up navigation around local schemes in general. But maybe I'm just doing it wrong in this patch.

@estark37, do you or any of your editors know the answer to my

> I did not understand the intent of https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-delivery-implicit-workers.

?

I'd like to keep moving on this if we can.

estark37 commented:
Unfortunately I don't understand that workers section at all either. @jeisinger do you know what's going on there by any chance (https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-delivery-implicit-workers)?

My best guess is that the intention was for workers to have the default policy (`no-referrer-when-downgrade`), and if there is a Referrer-Policy header in the worker's main script response, then that referrer policy should be applied for the worker. Does that sound at all reasonable?

domenic commented:
> My best guess is that the intention was for workers to have the default policy (no-referrer-when-downgrade), and if there is a Referrer-Policy header in the worker's main script response, then that referrer policy should be applied for the worker. Does that sound at all reasonable?

This sounds somewhat reasonable, but let me propose an alternative semantics:

- Fetching the worker itself uses the fetching client's referrer policy. (I.e., if fetching from a document, the one set by the header/meta tag for the document; if fetching from a worker, the one set in the next bullet point.)
- When the worker is fetched, check any Referrer-Policy header, and store it on the worker, so that future fetches from within that worker (importScripts, XHR, fetch, `import`, ...) all use that referrer policy by default.

I think we agree on bullet point two but IMO using the fetching client's referrer policy makes more sense than always using no-referrer-when-downgrade.

estark37 commented:
Ah, I was only thinking about fetches that the worker made. Using the
fetching client's policy when fetching the worker itself sgtm.

If there is no referrer policy header from fetching the worker, I think it
should probably just use the default for any fetches that the worker makes?

IIRC that matches how csp works: the fetching client's csp is applied when
fetching the worker itself, then the csp served with the worker script is
applied to fetches that the worker makes (and no csp if none was served
with the worker script)

On Thursday, May 12, 2016, Domenic Denicola <notifications@github.com>
wrote:

> My best guess is that the intention was for workers to have the default
> policy (no-referrer-when-downgrade), and if there is a Referrer-Policy
> header in the worker's main script response, then that referrer policy
> should be applied for the worker. Does that sound at all reasonable?
>
> This sounds somewhat reasonable, but let me propose an alternative
> semantics:
>
>    - Fetching the worker itself uses the fetching client's referrer
>    policy. (I.e., if fetching from a document, the one set by the header/meta
>    tag for the document; if fetching from a worker, the one set in the next
>    bullet point.)
>    - When the worker is fetched, check any Referrer-Policy header, and
>    store it on the worker, so that future fetches from within that worker
>    (importScripts, XHR, fetch, import, ...) all use that referrer policy
>    by default.
>
> I think we agree on bullet point two but IMO using the fetching client's
> referrer policy makes more sense than always using
> no-referrer-when-downgrade.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/whatwg/html/pull/1205#issuecomment-218947699>
>

estark37 commented:
(Sorry, I'm on my phone and don't think I was clear. I agree with your
proposal, Domenic, and just wanted to clarify what happens if there is no
referrer policy header served when fetching the worker itself)

On Thursday, May 12, 2016, Emily Stark <estark37@gmail.com> wrote:

> Ah, I was only thinking about fetches that the worker made. Using the
> fetching client's policy when fetching the worker itself sgtm.
>
> If there is no referrer policy header from fetching the worker, I think it
> should probably just use the default for any fetches that the worker makes?
>
> IIRC that matches how csp works: the fetching client's csp is applied when
> fetching the worker itself, then the csp served with the worker script is
> applied to fetches that the worker makes (and no csp if none was served
> with the worker script)
>
> On Thursday, May 12, 2016, Domenic Denicola <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> My best guess is that the intention was for workers to have the default
>> policy (no-referrer-when-downgrade), and if there is a Referrer-Policy
>> header in the worker's main script response, then that referrer policy
>> should be applied for the worker. Does that sound at all reasonable?
>>
>> This sounds somewhat reasonable, but let me propose an alternative
>> semantics:
>>
>>    - Fetching the worker itself uses the fetching client's referrer
>>    policy. (I.e., if fetching from a document, the one set by the header/meta
>>    tag for the document; if fetching from a worker, the one set in the next
>>    bullet point.)
>>    - When the worker is fetched, check any Referrer-Policy header, and
>>    store it on the worker, so that future fetches from within that worker
>>    (importScripts, XHR, fetch, import, ...) all use that referrer policy
>>    by default.
>>
>> I think we agree on bullet point two but IMO using the fetching client's
>> referrer policy makes more sense than always using
>> no-referrer-when-downgrade.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/whatwg/html/pull/1205#issuecomment-218947699>
>>
>

jeisinger commented:
The paragraph is supposed to define what happens if no referrer-policy header is in the response. In your two bullet points, it's then undefined which referrer policy to use, and the paragraph says that fetches from such an worker should use the default policy then.

domenic commented:
Sorry, it was implicit in my paragraphs, I agree. To be more clear: the worker referrer policy would stay at the  default of "", which in the "set request's referrer" algorithm, gets treated the same as no-referrer-when-downgrade, assuming there is no more specific override (e.g. via the fetch() API)

domenic commented:
I'll work on updating to reflect those bullet points and pong for another review when that's done. Not sure about the navigation stuff still, though. Maybe I'll try to see if we can copy CSP more closely or something...

domenic commented:
OK, it turns out I had already implemented those bullet points, so we should be good.

I looked at the navigation stuff more, and I think @bzbarsky is right that the local scheme division doesn't make sense. I think this is derived from [CSP's similar algorithm](https://w3c.github.io/webappsec-csp/#initialize-document-csp), but CSP says:

> Note: We do all this to ensure that a page cannot bypass its policy by embedding a frame or popping up a new window containing content it controls (blob: resources, or document.write()).

I don't believe this concern applies here; we don't need to "ratchet" the referrer policy to stay always as strict as the document's referrer policy. We can always override it by using `fetch` with a custom `referrerPolicy` option.

So I switched it to just always parsing the Referrer-Policy header, instead of only doing that for non-local schemes. Note that this doesn't quite mesh with my understanding of @bzbarsky's position:

>  If I have a document with referrer policy A which creates a subframe, then a document with referrer policy B navigates that subframe to a blob:, is the referrer policy of the result A or B? It should be B. I think this spec proposal makes it A.

I don't think we want to use the referrer policy of the document that initiated navigation as the referrer policy of the newly navigated-to document. Instead, I think we want to reset the referrer policy every navigation, based on the Referrer-Policy header (and falling back to "", which gives "no-referrer-if-downgrade" behavior). Thus, this is similar to the HTTPS state of the document.

---

So, this is ready for review, and IMO merge. Would love some sign-offs if people have the time.

bzbarsky commented:
The https state of the result loading a "blob:" is the https state of the thing that started the load.  I'm just saying referrer policy should work that way too...

Note that this is pretty specific to blob: and a few other schemes.  Obviously if an http:// URI is loaded, its referrer policy should come from headers.

domenic commented:
> The https state of the result loading a "blob:" is the https state of the thing that started the load. I'm just saying referrer policy should work that way too...

Hmm, is the spec currently wrong on that too? :( It currently says, in the navigate algorithm's "initialising a new Document object",

> Set the Document's HTTPS state to the HTTPS state of the resource used to generate the document.

domenic commented:
I guess maybe the complexity is hidden in the probably-underspecified process by which this "resource" is created...

bzbarsky commented:
>    Set the Document's HTTPS state to the HTTPS state of the resource used to generate the document.

Right, but the fetch spec propagates HTTPS state to blob responses.  See https://fetch.spec.whatwg.org/#basic-fetch the "blob:" part step 6.  Wish I could link to things more directly in that spec...

domenic commented:
Oh, thanks. It looks like we should do the same thing for referrer policy then.

bzbarsky commented:
Yes!   That's what I was trying to say in https://github.com/w3c/webappsec-referrer-policy/issues/32 but not clearly enough.

annevk commented:
See also https://github.com/whatwg/fetch/issues/270 about where blob URLs should grab HTTPS state from. The current thinking is that it should be stored on `Blob` objects directly.

bzbarsky commented:
Sure, but that leaves https state for about:blank anyway...

annevk commented on source:
```diff
@@ -8616,6 +8622,8 @@ interface <dfn>DOMStringMap</dfn> {
     a new <span data-x="concept-response">response</span> whose <span
     data-x="concept-response-body">body</span> is <var>source</var>, <span
+    data-x="concept-response-header-list">header list</span> is `<code
+    data-x="http-referrer-policy">Referrer-Policy</code>`/<var>referrer policy</var>, <span
```
Given that the document object created by the navigate algorithm has a "referrer policy", would it not be better to just set the policy directly there? Though I guess this also applies to CSP list and HTTPS state... Follow up?

annevk replied:
I'm saying this since we already have "When the navigate algorithm creates a Document object for this purpose, set that Document's reload override flag and set its reload override buffer to source." just after this sentence.

domenic replied:
The navigate algorithm sets the referrer policy, HTTPS state, and CSP list from the response that is (sort of) passed in to the navigate algorithm. So this seems like the right way to go.

The "reload override flag" and "reload override buffer" seem like things that are not part of a response, i.e. if we refactor the navigate algorithm, it will take a parameter list that looks something like (browsingContext, resource, exceptions enabled flag, reload override flag, reload override buffer). But referrer policy/CSP list/HTTPS state will be part of the response still.

(Not sure about "override URL"; it could go either way.)

annevk commented:
LGTM with that caveat.

At some point we should really refactor the document creation algorithms so that it's very clear what state gets copied across.
